toolchain/check/check.cpp

```diff
@@ -231,14 +250,360 @@ static auto InitPackageScopeAndImports(Context& context, UnitInfo& unit_info)
+              const auto& method =
+                  context.parse_tree().inline_methods().Get(method_index);
+              HandleFunctionDefinitionResume(context, method.start_id,
+                                             std::move(*sus_fn));
+              chunks.push_back(
+                  {.it = context.parse_tree().postorder(method.start_id).end(),
+                   .end = context.parse_tree()
+                              .postorder(method.definition_id)
+                              .end(),
+                   .next_method = next_inline_method.index});
+              ++method_index.index;
+              next_inline_method.SkipTo(method_index);
+            });
+        continue;
+      }
+
+      // If we're not parsing skipped methods, produce the next parse node for
+      // this chunk. If we've run out of parse nodes, we're done with this chunk
+      // of the parse tree.
+      if (chunks.back().it == chunks.back().end) {
+        auto old_chunk = chunks.pop_back_val();
+
+        // If we're out of chunks, then we're done entirely.
+        if (chunks.empty()) {
+          worklist.VerifyEmpty();
+          return std::nullopt;
+        }
+
+        next_inline_method.SkipTo(old_chunk.next_method);
+        continue;
+      }
+
+      auto node_id = *chunks.back().it;
+
+      // If we've reached the start of an inline method, skip to the end of it,
+      // and track that we need to parse it later.
+      if (node_id == next_inline_method.start_id) {
+        const auto& method =
+            context.parse_tree().inline_methods().Get(next_inline_method.index);
+        worklist.SuspendFunctionAndPush(context, next_inline_method.index,
+                                        method.start_id);
+
+        // Continue parsing after the end of the definition.
+        chunks.back().it =
+            context.parse_tree().postorder(method.definition_id).end();
+        next_inline_method.SkipTo(method.next_method_index);
+        continue;
+      }
+
+      ++chunks.back().it;
+      return node_id;
+    }
+  }
+
+  // Performs any processing necessary before handling a node.
+  auto BeforeHandle(Parse::NodeKind parse_kind) -> void {
+    // When we reach the start of a delayed method scope, add a task to the
+    // worklist to parse future skipped methods in the new context.
+    if (IsStartOfDelayedMethodScope(parse_kind)) {
+      worklist.PushEnterMethodScope(context);
+    }
+  }
+
+  // Performs any processing necessary after handling a node.
+  auto AfterHandle(Parse::NodeKind parse_kind) -> void {
+    // When we reach the end of a delayed method scope, add a task to the
+    // worklist to leave the scope. If this is not a nested scope, start parsing
+    // the skipped methods now.
+    if (IsEndOfDelayedMethodScope(parse_kind)) {
+      if (worklist.SuspendFinishedScopeAndPush(context)) {
+        chunks.back().parsing_skipped_methods = true;
+      }
+    }
```

Comment:
```suggestion
    if (IsEndOfDelayedMethodScope(parse_kind) &&
        worklist.SuspendFinishedScopeAndPush(context)) {
      chunks.back().parsing_skipped_methods = true;
    }
```

Mild preference for a single `if` where possible.

Reply:
Hm. I structured it like this because the `SuspendFinished...` call has important side-effects that should only happen if the `IsEnd...` call returns `true`, and I was concerned that `&&` wouldn't give a sufficient hint to the reader that the short-circuiting is important. Perhaps there's another structuring that would help here; how about this:

```suggestion
    if (IsEndOfDelayedMethodScope(parse_kind)) {
      chunks.back().parsing_skipped_methods =
          worklist.SuspendFinishedScopeAndPush(context);
    }
```

Reply:
SG